File: syndrome/core/disease.py
"""
Disease State Module

Implements disease state formalism:
- Disease vector D = (D_P, D_E, D_C, D_M, D_A, D_G, D_Ca, D_R)
- Disease classification by dominant component
- Disease severity computation
"""

from dataclasses import dataclass
from typing import List, Dict, Tuple, Optional
import numpy as np

from syndrome.core.coherence import Oscillator, coherence_index


# Oscillator class definitions
OSCILLATOR_CLASSES = ["P", "E", "C", "M", "A", "G", "Ca", "R"]

DISEASE_CLASS_NAMES = {
    "P": "Misfolding",
    "E": "Metabolic",
    "C": "Channelopathy",
    "M": "Excitability",
    "A": "Mitochondrial",
    "G": "Expression",
    "Ca": "Signaling",
    "R": "Rhythm",
}

DISEASE_CLASS_EXAMPLES = {
    "P": ["Alzheimer's", "Parkinson's", "Prion diseases", "Huntington's"],
    "E": ["Diabetes", "PKU", "Gaucher disease", "Galactosemia"],
    "C": ["Cystic fibrosis", "Long QT syndrome", "Myotonia", "Episodic ataxia"],
    "M": ["Epilepsy", "Cardiac arrhythmia", "Migraine"],
    "A": ["MELAS", "Leigh syndrome", "LHON", "Mitochondrial myopathy"],
    "G": ["Cancer", "Developmental disorders", "Thalassemia"],
    "Ca": ["Malignant hyperthermia", "Timothy syndrome", "Cardiac disorders"],
    "R": ["Sleep disorders", "Jet lag", "Seasonal affective disorder"],
}


@dataclass
class DiseaseVector:
    """
    Disease state vector across eight oscillator classes.

    D = (D_P, D_E, D_C, D_M, D_A, D_G, D_Ca, D_R)

    where D_i = 1 - η_i is the disease index for class i.
    """
    D_P: float  # Protein (misfolding)
    D_E: float  # Enzyme (metabolic)
    D_C: float  # Channel (channelopathy)
    D_M: float  # Membrane (excitability)
    D_A: float  # ATP (mitochondrial)
    D_G: float  # Genetic (expression)
    D_Ca: float  # Calcium (signaling)
    D_R: float  # Circadian (rhythm)

    def __post_init__(self) -> None:
        """Validate disease indices are in [0, 1]."""
        for name in OSCILLATOR_CLASSES:
            val = getattr(self, f"D_{name}")
            if not 0.0 <= val <= 1.0:
                raise ValueError(
                    f"Disease index D_{name} must be in [0, 1], got {val}"
                )

    def to_array(self) -> np.ndarray:
        """Convert to numpy array."""
        return np.array([
            self.D_P, self.D_E, self.D_C, self.D_M,
            self.D_A, self.D_G, self.D_Ca, self.D_R
        ], dtype=np.float64)

    def to_dict(self) -> Dict[str, float]:
        """Convert to dictionary."""
        return {
            "D_P": self.D_P,
            "D_E": self.D_E,
            "D_C": self.D_C,
            "D_M": self.D_M,
            "D_A": self.D_A,
            "D_G": self.D_G,
            "D_Ca": self.D_Ca,
            "D_R": self.D_R,
        }

    @classmethod
    def from_array(cls, arr: np.ndarray) -> "DiseaseVector":
        """Create from numpy array."""
        if len(arr) != 8:
            raise ValueError(f"Array must have length 8, got {len(arr)}")
        return cls(
            D_P=float(arr[0]),
            D_E=float(arr[1]),
            D_C=float(arr[2]),
            D_M=float(arr[3]),
            D_A=float(arr[4]),
            D_G=float(arr[5]),
            D_Ca=float(arr[6]),
            D_R=float(arr[7]),
        )

    @classmethod
    def from_coherences(
        cls,
        eta_P: float = 1.0,
        eta_E: float = 1.0,
        eta_C: float = 1.0,
        eta_M: float = 1.0,
        eta_A: float = 1.0,
        eta_G: float = 1.0,
        eta_Ca: float = 1.0,
        eta_R: float = 1.0,
    ) -> "DiseaseVector":
        """Create disease vector from coherence values."""
        return cls(
            D_P=1.0 - eta_P,
            D_E=1.0 - eta_E,
            D_C=1.0 - eta_C,
            D_M=1.0 - eta_M,
            D_A=1.0 - eta_A,
            D_G=1.0 - eta_G,
            D_Ca=1.0 - eta_Ca,
            D_R=1.0 - eta_R,
        )

    def dominant_class(self) -> str:
        """Get the dominant disease class (argmax D_i)."""
        arr = self.to_array()
        idx = int(np.argmax(arr))
        return OSCILLATOR_CLASSES[idx]

    def dominant_value(self) -> float:
        """Get the dominant disease index value."""
        return float(np.max(self.to_array()))

    def severity(self, weights: Optional[np.ndarray] = None) -> float:
        """
        Compute total disease severity.

        D_total = (1/W) Σᵢ wᵢ Dᵢ

        Args:
            weights: Optional weights for each class

        Returns:
            Disease severity in [0, 1]
        """
        arr = self.to_array()
        if weights is None:
            return float(np.mean(arr))
        else:
            weights = np.asarray(weights)
            return float(np.sum(weights * arr) / np.sum(weights))

    def classification(self) -> Tuple[str, str, List[str]]:
        """
        Full disease classification.

        Returns:
            Tuple of (class_code, class_name, example_diseases)
        """
        dom_class = self.dominant_class()
        return (
            dom_class,
            DISEASE_CLASS_NAMES[dom_class],
            DISEASE_CLASS_EXAMPLES[dom_class],
        )


def disease_vector(oscillators: List[Oscillator]) -> DiseaseVector:
    """
    Compute disease vector from oscillator ensemble.

    Groups oscillators by class and computes weighted disease index
    for each class.

    Args:
        oscillators: List of Oscillator objects

    Returns:
        DiseaseVector
    """
    # Group by class
    class_coherences: Dict[str, List[Tuple[float, float]]] = {
        c: [] for c in OSCILLATOR_CLASSES
    }

    for osc in oscillators:
        eta = osc.coherence()
        weight = osc.weight
        class_coherences[osc.osc_class].append((eta, weight))

    # Compute weighted average coherence per class
    disease_indices = {}
    for c in OSCILLATOR_CLASSES:
        values = class_coherences[c]
        if values:
            total_weight = sum(w for _, w in values)
            if total_weight > 0:
                weighted_eta = sum(eta * w for eta, w in values) / total_weight
            else:
                weighted_eta = 1.0  # Default to healthy
        else:
            weighted_eta = 1.0  # No oscillators = assume healthy

        disease_indices[f"D_{c}"] = 1.0 - weighted_eta

    return DiseaseVector(**disease_indices)


def disease_vector_from_dicts(oscillators: List[Dict]) -> DiseaseVector:
    """
    Compute disease vector from list of dictionaries.

    Args:
        oscillators: List of dicts with keys:
            'class', 'pi_obs', 'pi_opt', 'pi_deg', 'weight'

    Returns:
        DiseaseVector
    """
    osc_objects = [
        Oscillator(
            osc_class=d["class"],
            pi_obs=d["pi_obs"],
            pi_opt=d["pi_opt"],
            pi_deg=d["pi_deg"],
            weight=d.get("weight", 1.0),
        )
        for d in oscillators
    ]
    return disease_vector(osc_objects)


def classify_disease(D: DiseaseVector) -> str:
    """
    Classify disease by dominant component.

    class(D) = argmax_i D_i

    Args:
        D: Disease vector

    Returns:
        Disease class code (P, E, C, M, A, G, Ca, R)
    """
    return D.dominant_class()


def disease_severity(D: DiseaseVector, weights: Optional[np.ndarray] = None) -> float:
    """
    Compute total disease severity.

    Args:
        D: Disease vector
        weights: Optional class weights

    Returns:
        Severity in [0, 1]
    """
    return D.severity(weights)


def disease_signature_distance(D1: DiseaseVector, D2: DiseaseVector) -> float:
    """
    Compute distance between two disease signatures.

    Uses Euclidean distance in 8-dimensional disease space.

    Args:
        D1: First disease vector
        D2: Second disease vector

    Returns:
        Distance >= 0
    """
    return float(np.linalg.norm(D1.to_array() - D2.to_array()))


def healthy_vector() -> DiseaseVector:
    """Return a healthy disease vector (all zeros)."""
    return DiseaseVector(
        D_P=0.0, D_E=0.0, D_C=0.0, D_M=0.0,
        D_A=0.0, D_G=0.0, D_Ca=0.0, D_R=0.0
    )


def generate_disease_profile(
    dominant_class: str,
    severity: float,
    spread: float = 0.1
) -> DiseaseVector:
    """
    Generate a synthetic disease profile.

    Args:
        dominant_class: Dominant oscillator class
        severity: Severity of dominant class (0-1)
        spread: Random spread to other classes

    Returns:
        DiseaseVector with specified characteristics
    """
    if dominant_class not in OSCILLATOR_CLASSES:
        raise ValueError(f"Invalid class: {dominant_class}")

    indices = {}
    for c in OSCILLATOR_CLASSES:
        if c == dominant_class:
            indices[f"D_{c}"] = severity
        else:
            # Random background with given spread
            indices[f"D_{c}"] = np.random.uniform(0, spread)

    return DiseaseVector(**indices)

File: syndrome-rs/src/disease.rs
//! Disease State Module
//!
//! Implements disease state formalism:
//! - Disease vector D = (D_P, D_E, D_C, D_M, D_A, D_G, D_Ca, D_R)
//! - Disease classification by dominant component
//! - Disease severity computation

use serde::{Deserialize, Serialize};
use crate::coherence::{Oscillator, OscillatorClass};
use crate::{Result, SyndromeError};

/// Disease class names and descriptions.
pub const DISEASE_CLASS_NAMES: [&str; 8] = [
    "Misfolding",
    "Metabolic",
    "Channelopathy",
    "Excitability",
    "Mitochondrial",
    "Expression",
    "Signaling",
    "Rhythm",
];

/// Example diseases for each class.
pub const DISEASE_CLASS_EXAMPLES: [&[&str]; 8] = [
    &["Alzheimer's", "Parkinson's", "Prion diseases", "Huntington's"],
    &["Diabetes", "PKU", "Gaucher disease", "Galactosemia"],
    &["Cystic fibrosis", "Long QT syndrome", "Myotonia", "Episodic ataxia"],
    &["Epilepsy", "Cardiac arrhythmia", "Migraine"],
    &["MELAS", "Leigh syndrome", "LHON", "Mitochondrial myopathy"],
    &["Cancer", "Developmental disorders", "Thalassemia"],
    &["Malignant hyperthermia", "Timothy syndrome", "Cardiac disorders"],
    &["Sleep disorders", "Jet lag", "Seasonal affective disorder"],
];

/// Disease state vector across eight oscillator classes.
///
/// D = (D_P, D_E, D_C, D_M, D_A, D_G, D_Ca, D_R)
///
/// where D_i = 1 - η_i is the disease index for class i.
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub struct DiseaseVector {
    /// D_P: Protein (misfolding)
    pub d_p: f64,
    /// D_E: Enzyme (metabolic)
    pub d_e: f64,
    /// D_C: Channel (channelopathy)
    pub d_c: f64,
    /// D_M: Membrane (excitability)
    pub d_m: f64,
    /// D_A: ATP (mitochondrial)
    pub d_a: f64,
    /// D_G: Genetic (expression)
    pub d_g: f64,
    /// D_Ca: Calcium (signaling)
    pub d_ca: f64,
    /// D_R: Circadian (rhythm)
    pub d_r: f64,
}

impl DiseaseVector {
    /// Create a new disease vector with validation.
    ///
    /// # Errors
    /// Returns error if any component is outside [0, 1].
    pub fn new(
        d_p: f64,
        d_e: f64,
        d_c: f64,
        d_m: f64,
        d_a: f64,
        d_g: f64,
        d_ca: f64,
        d_r: f64,
    ) -> Result<Self> {
        let v = Self {
            d_p,
            d_e,
            d_c,
            d_m,
            d_a,
            d_g,
            d_ca,
            d_r,
        };
        v.validate()?;
        Ok(v)
    }

    /// Create without validation (for internal use).
    pub(crate) fn new_unchecked(
        d_p: f64,
        d_e: f64,
        d_c: f64,
        d_m: f64,
        d_a: f64,
        d_g: f64,
        d_ca: f64,
        d_r: f64,
    ) -> Self {
        Self {
            d_p,
            d_e,
            d_c,
            d_m,
            d_a,
            d_g,
            d_ca,
            d_r,
        }
    }

    /// Validate all components are in [0, 1].
    fn validate(&self) -> Result<()> {
        let components = [
            ("P", self.d_p),
            ("E", self.d_e),
            ("C", self.d_c),
            ("M", self.d_m),
            ("A", self.d_a),
            ("G", self.d_g),
            ("Ca", self.d_ca),
            ("R", self.d_r),
        ];

        for &(class, value) in &components {
            if value < 0.0 || value > 1.0 {
                return Err(SyndromeError::InvalidDiseaseIndex { class, value });
            }
        }
        Ok(())
    }

    /// Convert to array representation.
    #[inline]
    pub fn to_array(&self) -> [f64; 8] {
        [
            self.d_p, self.d_e, self.d_c, self.d_m,
            self.d_a, self.d_g, self.d_ca, self.d_r,
        ]
    }

    /// Create from array.
    ///
    /// # Errors
    /// Returns error if any value is outside [0, 1].
    pub fn from_array(arr: [f64; 8]) -> Result<Self> {
        Self::new(
            arr[0], arr[1], arr[2], arr[3],
            arr[4], arr[5], arr[6], arr[7],
        )
    }

    /// Create from coherence values.
    pub fn from_coherences(
        eta_p: f64,
        eta_e: f64,
        eta_c: f64,
        eta_m: f64,
        eta_a: f64,
        eta_g: f64,
        eta_ca: f64,
        eta_r: f64,
    ) -> Result<Self> {
        Self::new(
            1.0 - eta_p,
            1.0 - eta_e,
            1.0 - eta_c,
            1.0 - eta_m,
            1.0 - eta_a,
            1.0 - eta_g,
            1.0 - eta_ca,
            1.0 - eta_r,
        )
    }

    /// Get the dominant disease class (argmax D_i).
    pub fn dominant_class(&self) -> OscillatorClass {
        let arr = self.to_array();
        let (idx, _) = arr
            .iter()
            .enumerate()
            .max_by(|(_, a), (_, b)| a.partial_cmp(b).unwrap())
            .unwrap();
        OscillatorClass::from_index(idx).unwrap()
    }

    /// Get the dominant disease index value.
    pub fn dominant_value(&self) -> f64 {
        self.to_array()
            .iter()
            .cloned()
            .fold(f64::NEG_INFINITY, f64::max)
    }

    /// Compute total disease severity.
    ///
    /// D_total = (1/W) Σᵢ wᵢ Dᵢ
    ///
    /// If no weights provided, uses equal weights (mean).
    pub fn severity(&self, weights: Option<&[f64; 8]>) -> f64 {
        let arr = self.to_array();

        match weights {
            None => arr.iter().sum::<f64>() / 8.0,
            Some(w) => {
                let total_weight: f64 = w.iter().sum();
                if total_weight.abs() < 1e-10 {
                    return 0.0;
                }
                let weighted_sum: f64 = arr.iter().zip(w.iter()).map(|(d, w)| d * w).sum();
                weighted_sum / total_weight
            }
        }
    }

    /// Get index for given class.
    pub fn get(&self, class: OscillatorClass) -> f64 {
        match class {
            OscillatorClass::Protein => self.d_p,
            OscillatorClass::Enzyme => self.d_e,
            OscillatorClass::Channel => self.d_c,
            OscillatorClass::Membrane => self.d_m,
            OscillatorClass::Atp => self.d_a,
            OscillatorClass::Genetic => self.d_g,
            OscillatorClass::Calcium => self.d_ca,
            OscillatorClass::Circadian => self.d_r,
        }
    }

    /// Full disease classification.
    ///
    /// Returns (class_code, class_name, example_diseases).
    pub fn classification(&self) -> (&'static str, &'static str, &'static [&'static str]) {
        let class = self.dominant_class();
        let idx = class.index();
        (class.code(), DISEASE_CLASS_NAMES[idx], DISEASE_CLASS_EXAMPLES[idx])
    }
}

impl Default for DiseaseVector {
    /// Returns a healthy vector (all zeros).
    fn default() -> Self {
        Self::new_unchecked(0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0)
    }
}

/// Compute disease vector from oscillator ensemble.
///
/// Groups oscillators by class and computes weighted disease index
/// for each class.
pub fn disease_vector_from_oscillators(oscillators: &[Oscillator]) -> DiseaseVector {
    // Group by class and compute weighted average coherence
    let mut class_data: [Vec<(f64, f64)>; 8] = Default::default();

    for osc in oscillators {
        let eta = osc.coherence();
        let weight = osc.weight;
        class_data[osc.class.index()].push((eta, weight));
    }

    let mut disease_indices = [0.0; 8];
    for (i, values) in class_data.iter().enumerate() {
        if values.is_empty() {
            disease_indices[i] = 0.0; // Default to healthy
        } else {
            let total_weight: f64 = values.iter().map(|(_, w)| w).sum();
            if total_weight.abs() < 1e-10 {
                disease_indices[i] = 0.0;
            } else {
                let weighted_eta: f64 = values.iter().map(|(eta, w)| eta * w).sum::<f64>() / total_weight;
                disease_indices[i] = 1.0 - weighted_eta;
            }
        }
    }

    DiseaseVector::new_unchecked(
        disease_indices[0],
        disease_indices[1],
        disease_indices[2],
        disease_indices[3],
        disease_indices[4],
        disease_indices[5],
        disease_indices[6],
        disease_indices[7],
    )
}

/// Classify disease by dominant component.
///
/// class(D) = argmax_i D_i
pub fn classify_disease(d: &DiseaseVector) -> OscillatorClass {
    d.dominant_class()
}

/// Compute total disease severity.
pub fn disease_severity(d: &DiseaseVector, weights: Option<&[f64; 8]>) -> f64 {
    d.severity(weights)
}

/// Compute distance between two disease signatures.
///
/// Uses Euclidean distance in 8-dimensional disease space.
pub fn disease_signature_distance(d1: &DiseaseVector, d2: &DiseaseVector) -> f64 {
    let a1 = d1.to_array();
    let a2 = d2.to_array();

    let sum_sq: f64 = a1
        .iter()
        .zip(a2.iter())
        .map(|(x, y)| (x - y).powi(2))
        .sum();

    sum_sq.sqrt()
}

/// Return a healthy disease vector (all zeros).
pub fn healthy_vector() -> DiseaseVector {
    DiseaseVector::default()
}

/// Generate a synthetic disease profile.
pub fn generate_disease_profile(
    dominant_class: OscillatorClass,
    severity: f64,
    spread: f64,
) -> DiseaseVector {
    use rand::Rng;
    let mut rng = rand::thread_rng();

    let mut indices = [0.0; 8];
    let dom_idx = dominant_class.index();

    for (i, val) in indices.iter_mut().enumerate() {
        if i == dom_idx {
            *val = severity.clamp(0.0, 1.0);
        } else {
            *val = (rng.gen::<f64>() * spread).clamp(0.0, 1.0);
        }
    }

    DiseaseVector::new_unchecked(
        indices[0], indices[1], indices[2], indices[3],
        indices[4], indices[5], indices[6], indices[7],
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_healthy_vector() {
        let h = healthy_vector();
        assert!(h.severity(None).abs() < 1e-10);
    }

    #[test]
    fn test_disease_vector_validation() {
        assert!(DiseaseVector::new(0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5).is_ok());
        assert!(DiseaseVector::new(1.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5).is_err());
        assert!(DiseaseVector::new(-0.1, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5).is_err());
    }

    #[test]
    fn test_dominant_class() {
        let d = DiseaseVector::new(0.8, 0.2, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1).unwrap();
        assert_eq!(d.dominant_class(), OscillatorClass::Protein);

        let d = DiseaseVector::new(0.1, 0.9, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1).unwrap();
        assert_eq!(d.dominant_class(), OscillatorClass::Enzyme);
    }

    #[test]
    fn test_severity() {
        let d = DiseaseVector::new(0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5).unwrap();
        assert!((d.severity(None) - 0.5).abs() < 1e-10);

        let d = DiseaseVector::new(1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0).unwrap();
        assert!((d.severity(None) - 0.125).abs() < 1e-10);
    }

    #[test]
    fn test_from_coherences() {
        let d = DiseaseVector::from_coherences(0.8, 0.9, 0.7, 0.85, 0.6, 0.95, 0.75, 0.9).unwrap();
        assert!((d.d_p - 0.2).abs() < 1e-10);
        assert!((d.d_e - 0.1).abs() < 1e-10);
    }

    #[test]
    fn test_disease_signature_distance() {
        let d1 = healthy_vector();
        let d2 = healthy_vector();
        assert!(disease_signature_distance(&d1, &d2).abs() < 1e-10);

        let d1 = DiseaseVector::new(1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0).unwrap();
        let d2 = DiseaseVector::new(0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0).unwrap();
        assert!((disease_signature_distance(&d1, &d2) - 2.0_f64.sqrt()).abs() < 1e-10);
    }

    #[test]
    fn test_disease_from_oscillators() {
        let oscillators = vec![
            Oscillator::new(OscillatorClass::Protein, 0.2, 1.0, 0.0),
            Oscillator::new(OscillatorClass::Enzyme, 0.8, 1.0, 0.0),
        ];
        let d = disease_vector_from_oscillators(&oscillators);
        assert!((d.d_p - 0.8).abs() < 1e-10);
        assert!((d.d_e - 0.2).abs() < 1e-10);
    }

    #[test]
    fn test_classification() {
        let d = DiseaseVector::new(0.85, 0.2, 0.15, 0.3, 0.25, 0.18, 0.22, 0.12).unwrap();
        let (code, name, _) = d.classification();
        assert_eq!(code, "P");
        assert_eq!(name, "Misfolding");
    }
}
